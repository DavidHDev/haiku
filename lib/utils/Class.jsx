import React from "react";

export const Class = ({ className, toggleClass, isActive, children }) => {
  const classes = `${className} ${isActive ? toggleClass : ""}`;
  return <div className={classes}>{children}</div>;
};
