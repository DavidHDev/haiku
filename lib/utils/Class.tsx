import React, { ReactNode } from 'react';

interface ClassProps {
  className?: string;
  condition?: boolean;
  toggleClass?: string;
  children: ReactNode;
}

/**
 * Class component that conditionally applies a class name to a div element.
 *
 * @param className - The initial class name for the div element.
 * @param condition - The condition to determine whether to apply the toggle class or not.
 * @param toggleClass - The class name to be toggled based on the condition.
 * @param children - The content to be rendered inside the div element.
 * @returns The rendered Class component.
 */
export const Class: React.FC<ClassProps> = ({
  className = '',
  condition = false,
  toggleClass = '',
  children,
}) => {
  const computedClassName = condition
    ? `${className} ${toggleClass}`.trim()
    : className;

  return <div className={computedClassName}>{children}</div>;
};
