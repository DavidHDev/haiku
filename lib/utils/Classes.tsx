import React, { ReactNode, ElementType } from 'react';

// Check if a value is a plain object literal
const isObjectLiteral = (value: any): boolean => {
  return Object.prototype.toString.call(value) === '[object Object]';
};

// Remove duplicate class names from a string
const deduplicateClasses = (classes: string) =>
  [...new Set(classes.split(' '))].join(' ');

interface ClassProps {
  className?: string;
  toggleClasses?: Record<string, boolean>;
  children: ReactNode;
  as?: ElementType;
  [key: string]: any;
}

/**
 * Classes component that conditionally applies class names to a specified HTML element.
 *
 * @param className - The initial class name for the element.
 * @param toggleClasses - An object with condition-to-classes mappings.
 * @param children - The content to be rendered inside the specified HTML element.
 * @param as - The type of HTML element to render.
 * @param props - Any additional props to be passed to the element.
 * @returns The rendered Class component.
 */

export const Class: React.FC<ClassProps> = ({
  className = '',
  toggleClasses = {},
  children,
  as: Component = 'div',
  ...props
}) => {
  if (!isObjectLiteral(toggleClasses)) {
    console.error(
      'toggleClasses prop must be an object literal. Received:',
      toggleClasses,
    );

    return (
      <Component className={className} {...props}>
        {children}
      </Component>
    );
  }

  let computedClassNames = className;

  for (const [classes, condition] of Object.entries(toggleClasses)) {
    if (condition) {
      computedClassNames += ` ${classes.trim()}`;
    }
  }

  computedClassNames = deduplicateClasses(computedClassNames);

  return (
    <Component className={computedClassNames} {...props}>
      {children}
    </Component>
  );
};
