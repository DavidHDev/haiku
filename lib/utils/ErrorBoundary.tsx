import React from 'react';

interface ErrorBoundaryProps {
  children: React.ReactNode;
  fallback: React.ComponentType<{
    error: Error | null;
    errorInfo: React.ErrorInfo | null;
    retry: () => void;
  }>;
}

interface ErrorBoundaryState {
  hasError: boolean;
  error: Error | null;
  errorInfo: React.ErrorInfo | null;
}

class ErrorBoundary extends React.Component<
  ErrorBoundaryProps,
  ErrorBoundaryState
> {
  constructor(props: ErrorBoundaryProps) {
    super(props);
    this.state = {
      hasError: false,
      error: null,
      errorInfo: null,
    };
  }

  static getDerivedStateFromError(error: Error): Partial<ErrorBoundaryState> {
    return {
      hasError: true,
      error,
    };
  }

  componentDidCatch(error: Error, errorInfo: React.ErrorInfo) {
    this.setState({
      error: error,
      errorInfo: errorInfo,
    });
  }

  render() {
    if (this.state.hasError) {
      const Fallback = this.props.fallback;
      return (
        <Fallback
          error={this.state.error}
          errorInfo={this.state.errorInfo}
          retry={() => this.setState({ hasError: false })}
        />
      );
    }

    return this.props.children;
  }
}

export default ErrorBoundary;
