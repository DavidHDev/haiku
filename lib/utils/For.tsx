import { Children, type ReactNode } from 'react';

type RenderFn = <T>(item: T, index?: number) => ReactNode;

export const For = <T,>({ render, each }: { render: RenderFn; each: T[] }) =>
  Children.toArray(each.map((item, index) => render(item, index)));
