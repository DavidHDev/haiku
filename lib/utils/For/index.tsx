import { Children } from 'react'

interface IProps<T> {
    each: T[]
    render: (data: T, index: number) => React.ReactNode
}

export function For<T>({ each, render }: IProps<T>): JSX.Element {
    // @todo `key={}`

    return (
        <>{Children.toArray(each.map((item, index) => render(item, index)))}</>
    )
}
