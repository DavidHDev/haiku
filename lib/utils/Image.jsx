import React from "react";

export const Image = ({
  src,
  loading,
  alt,
  fallback,
  style,
  width,
  height,
  className,
}) => {
  const handleBrokenImage = (event) => (event.target.src = fallback);

  return (
    <img
      src={src}
      loading={loading}
      alt={alt}
      onError={handleBrokenImage}
      style={style}
      width={width}
      height={height}
      className={className}
    />
  );
};
