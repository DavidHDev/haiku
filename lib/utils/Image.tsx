import { ImgHTMLAttributes } from 'react';

type Props = ImgHTMLAttributes<HTMLImageElement> & { fallback: string };

export const Image = ({ src, loading, alt, fallback, ...props }: Props) => {
  const handleBrokenImage = (event: any) => (event.target.src = fallback);

  return (
    <img
      src={src}
      loading={loading}
      alt={alt}
      onError={handleBrokenImage}
      {...props}
    />
  );
};
