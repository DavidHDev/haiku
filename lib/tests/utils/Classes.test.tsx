import { render, screen } from '@testing-library/react';
import { Classes } from '../../utils/Classes';

describe('Classes component', () => {
  const text = 'Classes Block';

  it('renders the Classes component', () => {
    const { asFragment } = render(
      <Classes className="base-class">{text}</Classes>,
    );
    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByText(text)).toBeInTheDocument();
  });

  it('renders with base className', () => {
    render(<Classes className="base-class">{text}</Classes>);

    const element = screen.queryByText(text);
    expect(element).toHaveClass('base-class');
  });

  it('applies toggleClasses when condition is true', () => {
    render(
      <Classes
        as="section"
        className="base-class"
        toggleClasses={{ 'active-class': true }}
      >
        {text}
      </Classes>,
    );
    const element = screen.getByText(text);
    expect(element).toHaveClass('base-class active-class');
  });

  it('does not apply toggleClasses when condition is false', () => {
    render(
      <Classes
        className="base-class"
        toggleClasses={{ 'inactive-class': false }}
      >
        {text}
      </Classes>,
    );
    const element = screen.getByText(text);
    expect(element).not.toHaveClass('inactive-class');
  });

  it('removes duplicate classes', () => {
    render(
      <Classes
        className="base-class"
        toggleClasses={{ 'active-class focus base-class': true }}
      >
        {text}
      </Classes>,
    );
    const element = screen.getByText(text);
    expect(element).toHaveClass('base-class active-class focus');
  });
});
