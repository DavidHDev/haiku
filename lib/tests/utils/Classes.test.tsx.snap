// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Classes component renders the Classes component 1`] = `
<DocumentFragment>
  <div
    class="base-class"
  >
    Classes Block
  </div>
</DocumentFragment>
`;
